File: nxflix-agents-ts/src/providers/llm.ts
import { generateText, generateObject } from 'ai';
import { createOpenAI } from '@ai-sdk/openai';
import { createAnthropic } from '@ai-sdk/anthropic';
import { createGoogleGenerativeAI } from '@ai-sdk/google';
import { createOllama } from 'ollama-ai-provider';
import { z, ZodType } from 'zod';
import { settings } from '../config.js';
import { createTrace } from '../tracing/index.js';

export interface Message {
  role: 'system' | 'user' | 'assistant';
  content: string;
}

export interface CompletionOptions {
  temperature?: number;
  maxTokens?: number;
}

type Provider = 'openai' | 'anthropic' | 'gemini' | 'ollama';

function getProvider(provider: Provider) {
  switch (provider) {
    case 'openai':
      return createOpenAI({ apiKey: settings.openaiApiKey });
    case 'anthropic':
      return createAnthropic({ apiKey: settings.anthropicApiKey });
    case 'gemini':
      return createGoogleGenerativeAI({ apiKey: settings.googleApiKey });
    case 'ollama':
      return createOllama({ baseURL: settings.ollamaBaseUrl });
  }
}

export class LLMProvider {
  private provider: Provider;
  private model: string;

  constructor(provider?: Provider, model?: string) {
    this.provider = provider ?? settings.defaultProvider;
    this.model = model ?? settings.defaultModel;
  }

  async complete(
    messages: Message[],
    options: CompletionOptions = {}
  ): Promise<string> {
    const { temperature = 0.7, maxTokens = 1024 } = options;

    const trace = createTrace('llm.complete', {
      provider: this.provider,
      model: this.model,
      messageCount: messages.length,
    });

    try {
      const provider = getProvider(this.provider);
      const result = await generateText({
        model: provider(this.model),
        messages: messages.map((m) => ({
          role: m.role,
          content: m.content,
        })),
        temperature,
        maxTokens,
      });

      trace?.update({ output: { textLength: result.text.length } });
      trace?.end();

      return result.text;
    } catch (error) {
      trace?.update({ output: { error: String(error) } });
      trace?.end();
      throw error;
    }
  }

  async completeJson<T>(
    messages: Message[],
    schema: ZodType<T>,
    options: CompletionOptions = {}
  ): Promise<T> {
    const { temperature = 0.3, maxTokens = 2048 } = options;

    const trace = createTrace('llm.completeJson', {
      provider: this.provider,
      model: this.model,
      messageCount: messages.length,
    });

    try {
      const provider = getProvider(this.provider);
      const result = await generateObject({
        model: provider(this.model),
        messages: messages.map((m) => ({
          role: m.role,
          content: m.content,
        })),
        schema,
        temperature,
        maxTokens,
      });

      trace?.update({ output: { success: true } });
      trace?.end();

      return result.object;
    } catch (error) {
      trace?.update({ output: { error: String(error) } });
      trace?.end();
      throw error;
    }
  }
}

export async function complete(
  messages: Message[],
  provider?: Provider,
  model?: string,
  options?: CompletionOptions
): Promise<string> {
  const llm = new LLMProvider(provider, model);
  return llm.complete(messages, options);
}

export async function completeWithJson<T>(
  messages: Message[],
  schema: ZodType<T>,
  provider?: Provider,
  model?: string,
  options?: CompletionOptions
): Promise<T> {
  const llm = new LLMProvider(provider, model);
  return llm.completeJson(messages, schema, options);
}

File: nxflix-agents-python/src/nxflix_agents/providers/llm.py
"""LiteLLM-based multi-provider LLM abstraction."""

import json
from typing import Any, TypeVar

from pydantic import BaseModel

from nxflix_agents.config import get_settings
from nxflix_agents.tracing import get_opik_logger

T = TypeVar("T", bound=BaseModel)


class LLMProvider:
    """Multi-provider LLM client using LiteLLM."""

    def __init__(
        self,
        provider: str | None = None,
        model: str | None = None,
    ):
        settings = get_settings()
        self.provider = provider or settings.default_provider
        self.model = model or settings.default_model
        self._model_string = settings.get_model_string(self.provider, self.model)

    async def complete(
        self,
        messages: list[dict[str, str]],
        temperature: float = 0.7,
        max_tokens: int = 1024,
        **kwargs: Any,
    ) -> str:
        """Generate a completion from the LLM."""
        from litellm import acompletion

        settings = get_settings()
        callbacks = []

        opik_logger = get_opik_logger()
        if opik_logger:
            callbacks.append(opik_logger)

        # Set provider-specific API keys
        api_key = None
        api_base = None

        if self.provider == "openai":
            api_key = settings.openai_api_key
        elif self.provider == "anthropic":
            api_key = settings.anthropic_api_key
        elif self.provider == "gemini":
            api_key = settings.google_api_key
        elif self.provider == "ollama":
            api_base = settings.ollama_base_url

        response = await acompletion(
            model=self._model_string,
            messages=messages,
            temperature=temperature,
            max_tokens=max_tokens,
            api_key=api_key,
            api_base=api_base,
            callbacks=callbacks if callbacks else None,
            **kwargs,
        )

        return response.choices[0].message.content or ""

    async def complete_json(
        self,
        messages: list[dict[str, str]],
        response_model: type[T],
        temperature: float = 0.3,
        max_tokens: int = 2048,
        **kwargs: Any,
    ) -> T:
        """Generate a structured JSON response from the LLM."""
        schema_json = json.dumps(response_model.model_json_schema(), indent=2)
        system_message = (
            "You must respond with valid JSON that matches this schema:\n"
            f"```json\n{schema_json}\n```\n"
            "Do not include any text before or after the JSON."
        )

        # Prepend system message for JSON mode
        messages_with_system = [{"role": "system", "content": system_message}] + messages

        response = await self.complete(
            messages=messages_with_system,
            temperature=temperature,
            max_tokens=max_tokens,
            **kwargs,
        )

        # Parse and validate response
        try:
            # Handle markdown code blocks
            content = response.strip()
            if content.startswith("```"):
                lines = content.split("\n")
                content = "\n".join(lines[1:-1])

            data = json.loads(content)
            return response_model.model_validate(data)
        except (json.JSONDecodeError, ValueError) as e:
            raise ValueError(f"Failed to parse LLM response as {response_model.__name__}: {e}")


# Convenience functions
async def complete(
    messages: list[dict[str, str]],
    provider: str | None = None,
    model: str | None = None,
    **kwargs: Any,
) -> str:
    """Generate a completion using the default or specified provider."""
    llm = LLMProvider(provider=provider, model=model)
    return await llm.complete(messages, **kwargs)


async def complete_with_json(
    messages: list[dict[str, str]],
    response_model: type[T],
    provider: str | None = None,
    model: str | None = None,
    **kwargs: Any,
) -> T:
    """Generate a structured JSON response using the default or specified provider."""
    llm = LLMProvider(provider=provider, model=model)
    return await llm.complete_json(messages, response_model, **kwargs)
